File: python/src/data_platform_lab/sensor_demo.py
"""Sensor pipeline demo — ingest, validate, deduplicate, aggregate, and output.

Reads ``data/sample/sensor_events.json`` (JSONL), runs five orchestrated
steps through the :class:`Pipeline` runner with observability tracking,
dead-letter routing, hourly aggregation, and a manifest.

Run from the repository root::

    cd python
    poetry run python -m data_platform_lab.sensor_demo

Or with custom directories::

    poetry run python -m data_platform_lab.sensor_demo --data-dir ../data/sample
"""

from __future__ import annotations

import csv
import json
import logging
from datetime import datetime
from pathlib import Path
from typing import Any

from data_platform_lab.manifest import write_manifest
from data_platform_lab.observability.tracker import (
    RunTracker,
    format_run_metadata,
)
from data_platform_lab.orchestration.runner import Pipeline, format_result

logger = logging.getLogger(__name__)

REQUIRED_FIELDS = ("sensor_id", "type", "value", "unit", "location", "timestamp")


# ---------------------------------------------------------------------------
# Pipeline steps
# ---------------------------------------------------------------------------


def _ingest(ctx: dict[str, Any]) -> dict[str, Any]:
    """Step 1: read sensor_events.json JSONL and parse each line."""
    data_dir = Path(ctx["data_dir"])
    path = data_dir / "sensor_events.json"

    raw_events: list[dict[str, Any]] = []
    with path.open(encoding="utf-8") as fh:
        for line in fh:
            line = line.strip()
            if line:
                raw_events.append(json.loads(line))

    ctx["raw_events"] = raw_events
    ctx["tracker"].inc_files_processed()
    ctx["tracker"].inc_rows_read(len(raw_events))

    return {"events_read": len(raw_events)}


def _validate(ctx: dict[str, Any]) -> dict[str, Any]:
    """Step 2: validate each event; split into accepted and rejected."""
    accepted: list[dict[str, Any]] = []
    rejected: list[dict[str, Any]] = []
    rejection_reasons: dict[str, int] = {}

    for event in ctx["raw_events"]:
        reasons: list[str] = []

        # Check required fields
        for field in REQUIRED_FIELDS:
            if field not in event:
                reasons.append(f"missing_{field}")

        # Value must be a number (not null / None)
        if "value" in event and not isinstance(event["value"], (int, float)):
            reasons.append("value_not_numeric")

        # Timestamp must be parseable
        if "timestamp" in event:
            try:
                datetime.fromisoformat(event["timestamp"].replace("Z", "+00:00"))
            except (ValueError, AttributeError):
                reasons.append("invalid_timestamp")

        if reasons:
            rejected.append({**event, "_rejection_reasons": reasons})
            for r in reasons:
                rejection_reasons[r] = rejection_reasons.get(r, 0) + 1
        else:
            accepted.append(event)

    ctx["accepted"] = accepted
    ctx["rejected"] = rejected
    ctx["tracker"].inc_rows_rejected(len(rejected))

    return {
        "accepted": len(accepted),
        "rejected": len(rejected),
        "rejection_reasons": rejection_reasons,
    }


def _deduplicate(ctx: dict[str, Any]) -> dict[str, Any]:
    """Step 3: deduplicate by sensor_id::timestamp key (first wins)."""
    seen: set[str] = set()
    deduplicated: list[dict[str, Any]] = []
    duplicates: list[dict[str, Any]] = []

    for event in ctx["accepted"]:
        key = f"{event['sensor_id']}::{event['timestamp']}"
        if key in seen:
            duplicates.append(event)
        else:
            seen.add(key)
            deduplicated.append(event)

    ctx["deduplicated"] = deduplicated
    ctx["duplicates"] = duplicates

    return {
        "before": len(ctx["accepted"]),
        "after": len(deduplicated),
        "duplicates_removed": len(duplicates),
    }


def _aggregate(ctx: dict[str, Any]) -> dict[str, Any]:
    """Step 4: compute hourly aggregates by sensor_id and per-location totals."""
    hourly: dict[str, list[float]] = {}
    location_events: dict[str, int] = {}
    location_sensors: dict[str, set[str]] = {}

    for event in ctx["deduplicated"]:
        ts = datetime.fromisoformat(event["timestamp"].replace("Z", "+00:00"))
        hour_key = ts.strftime("%Y-%m-%dT%H:00:00Z")
        bucket_key = f"{event['sensor_id']}|{hour_key}"

        hourly.setdefault(bucket_key, []).append(float(event["value"]))

        loc = event["location"]
        location_events[loc] = location_events.get(loc, 0) + 1
        location_sensors.setdefault(loc, set()).add(event["sensor_id"])

    hourly_aggregates: list[dict[str, Any]] = []
    for bucket_key, values in sorted(hourly.items()):
        sensor_id, hour = bucket_key.split("|", 1)
        hourly_aggregates.append({
            "sensor_id": sensor_id,
            "hour": hour,
            "count": len(values),
            "min": min(values),
            "max": max(values),
            "avg": round(sum(values) / len(values), 4),
        })

    location_summary: list[dict[str, Any]] = []
    for loc in sorted(location_events):
        location_summary.append({
            "location": loc,
            "event_count": location_events[loc],
            "sensor_count": len(location_sensors[loc]),
        })

    ctx["hourly_aggregates"] = hourly_aggregates
    ctx["location_summary"] = location_summary

    return {
        "hourly_buckets": len(hourly_aggregates),
        "locations": len(location_summary),
    }


def _output(ctx: dict[str, Any]) -> dict[str, Any]:
    """Step 5: write output files."""
    output_dir = Path(ctx["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)
    tracker: RunTracker = ctx["tracker"]

    # accepted.jsonl — valid deduplicated events
    accepted_path = output_dir / "accepted.jsonl"
    with accepted_path.open("w", encoding="utf-8") as fh:
        for event in ctx["deduplicated"]:
            fh.write(json.dumps(event) + "\n")
    tracker.inc_rows_written(len(ctx["deduplicated"]))

    # dead_letter.jsonl — rejected + duplicate events with reason
    dead_letter_path = output_dir / "dead_letter.jsonl"
    with dead_letter_path.open("w", encoding="utf-8") as fh:
        for event in ctx["rejected"]:
            fh.write(json.dumps(event) + "\n")
        for event in ctx["duplicates"]:
            fh.write(json.dumps({**event, "_rejection_reasons": ["duplicate"]}) + "\n")

    # hourly_aggregates.csv
    agg_path = output_dir / "hourly_aggregates.csv"
    with agg_path.open("w", newline="", encoding="utf-8") as fh:
        writer = csv.DictWriter(
            fh, fieldnames=["sensor_id", "hour", "count", "min", "max", "avg"],
        )
        writer.writeheader()
        writer.writerows(ctx["hourly_aggregates"])

    # location_summary.csv
    loc_path = output_dir / "location_summary.csv"
    with loc_path.open("w", newline="", encoding="utf-8") as fh:
        writer = csv.DictWriter(
            fh, fieldnames=["location", "event_count", "sensor_count"],
        )
        writer.writeheader()
        writer.writerows(ctx["location_summary"])

    # summary.json
    summary_path = output_dir / "summary.json"
    summary = {
        "pipeline_name": ctx["pipeline_name"],
        "events_read": len(ctx["raw_events"]),
        "accepted": len(ctx["deduplicated"]),
        "rejected": len(ctx["rejected"]),
        "duplicates_removed": len(ctx["duplicates"]),
        "hourly_buckets": len(ctx["hourly_aggregates"]),
        "locations": len(ctx["location_summary"]),
        "step_results": ctx["step_results"],
    }
    summary_path.write_text(json.dumps(summary, indent=2), encoding="utf-8")

    return {"files_written": 5, "output_dir": str(output_dir)}


# ---------------------------------------------------------------------------
# Main pipeline function
# ---------------------------------------------------------------------------


def run_sensor_demo(
    data_dir: str | Path = "data/sample",
    output_dir: str | Path = "data/silver/sensor_demo",
    manifest_dir: str | Path = "data/manifests",
) -> dict[str, Any]:
    """Run the sensor pipeline demo.

    Returns a dict with:
    - ``pipeline_result``: the :class:`PipelineResult` from the orchestration runner
    - ``metadata``: :class:`RunMetadata` from the tracker
    - ``manifest_path``: path to the manifest file
    - ``output_dir``: output directory path
    """
    data_dir = Path(data_dir)
    output_dir = Path(output_dir)
    manifest_dir = Path(manifest_dir)

    tracker = RunTracker("sensor_demo")

    with tracker:
        # Build the shared context
        ctx: dict[str, Any] = {
            "data_dir": str(data_dir),
            "output_dir": str(output_dir),
            "tracker": tracker,
        }

        # Wire up the pipeline
        pipeline = Pipeline("sensor_demo")
        pipeline.add_step("ingest", _ingest)
        pipeline.add_step("validate", _validate)
        pipeline.add_step("deduplicate", _deduplicate)
        pipeline.add_step("aggregate", _aggregate)
        pipeline.add_step("output", _output)

        pipeline_result = pipeline.run(ctx)

        # Feed pipeline counters into the tracker
        tracker.set_extra("events_read", len(ctx.get("raw_events", [])))
        tracker.set_extra("accepted", len(ctx.get("deduplicated", [])))
        tracker.set_extra("rejected", len(ctx.get("rejected", [])))
        tracker.set_extra("duplicates_removed", len(ctx.get("duplicates", [])))
        tracker.set_extra("output_dir", str(output_dir))

    meta = tracker.metadata
    run_id = meta.run_id

    # Write manifest
    output_files = [
        str(output_dir / f)
        for f in (
            "accepted.jsonl",
            "dead_letter.jsonl",
            "hourly_aggregates.csv",
            "location_summary.csv",
            "summary.json",
        )
    ]
    manifest_path = write_manifest(
        pipeline_name="sensor_demo",
        run_id=run_id,
        source=str(data_dir / "sensor_events.json"),
        output=output_files,
        row_count=len(ctx.get("deduplicated", [])),
        status=pipeline_result.status,
        schema_hint=list(REQUIRED_FIELDS),
        warnings=meta.warnings or None,
        extras={
            "pipeline_duration_seconds": pipeline_result.duration_seconds,
            "steps_passed": pipeline_result.steps_passed,
            "steps_failed": pipeline_result.steps_failed,
        },
        manifest_dir=manifest_dir,
    )

    return {
        "pipeline_result": pipeline_result,
        "metadata": meta,
        "manifest_path": manifest_path,
        "output_dir": output_dir,
    }


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------


def main() -> None:
    """CLI entry point for the sensor demo."""
    import argparse

    parser = argparse.ArgumentParser(
        description="Run the sensor pipeline demo.",
    )
    parser.add_argument(
        "--data-dir",
        default="../data/sample",
        help="Directory containing sensor_events.json.",
    )
    parser.add_argument(
        "--output-dir",
        default="../data/silver/sensor_demo",
        help="Directory for output files.",
    )
    parser.add_argument(
        "--manifest-dir",
        default="../data/manifests",
        help="Directory for the run manifest JSON.",
    )
    args = parser.parse_args()

    logging.basicConfig(level=logging.INFO, format="%(levelname)s %(name)s: %(message)s")

    result = run_sensor_demo(args.data_dir, args.output_dir, args.manifest_dir)

    print()
    print(format_result(result["pipeline_result"]))
    print()
    print(format_run_metadata(result["metadata"]))
    print()
    print(f"Output: {result['output_dir']}")
    print(f"Manifest: {result['manifest_path']}")


if __name__ == "__main__":
    main()

File: javascript/src/sensor-demo.js
/**
 * Sensor pipeline demo — orchestrated 5-step workflow.
 *
 * Processes data/sample/sensor_events.json through ingest, validate,
 * deduplicate, aggregate, and output steps with observability,
 * dead-letter routing, hourly aggregation, and a manifest.
 *
 * Run from the javascript/ directory:
 *   node src/sensor-demo.js
 *
 * Or with custom paths:
 *   node src/sensor-demo.js --data-dir ../data/sample --output-dir ../data/silver/sensor_demo
 */

import { readFile, writeFile, mkdir } from "node:fs/promises";
import { join } from "node:path";
import { parseArgs } from "node:util";

import { Pipeline, formatResult } from "./orchestration/runner.js";
import { writeManifest, generateRunId } from "./manifest.js";

// ---------------------------------------------------------------------------
// Step functions
// ---------------------------------------------------------------------------

/**
 * Step 1: Ingest — read sensor_events.json (JSONL), parse each line.
 */
async function ingest(ctx) {
  const filePath = join(ctx.dataDir, "sensor_events.json");
  const content = await readFile(filePath, "utf-8");
  const lines = content
    .split(/\r?\n/)
    .filter((line) => line.trim().length > 0);

  const events = lines.map((line, idx) => {
    try {
      return JSON.parse(line);
    } catch {
      return { _parse_error: true, _line: idx + 1, _raw: line };
    }
  });

  ctx.raw_events = events;
  return { events_read: events.length };
}

/**
 * Step 2: Validate — check required fields, numeric value, parseable timestamp.
 * Split into accepted / rejected (dead-letter).
 */
function validate(ctx) {
  const required = ["sensor_id", "type", "value", "unit", "location", "timestamp"];
  const accepted = [];
  const rejected = [];

  for (const event of ctx.raw_events) {
    const reasons = [];

    // Parse errors from ingest
    if (event._parse_error) {
      reasons.push(`JSON parse error on line ${event._line}`);
      rejected.push({ event, reasons });
      continue;
    }

    // Required fields
    for (const field of required) {
      if (!(field in event) || event[field] === undefined) {
        reasons.push(`missing field: ${field}`);
      }
    }

    // Numeric value (null counts as invalid)
    if (event.value === null || event.value === undefined || typeof event.value !== "number" || !Number.isFinite(event.value)) {
      reasons.push("value is not a finite number");
    }

    // Parseable timestamp
    if (event.timestamp) {
      const ts = Date.parse(event.timestamp);
      if (Number.isNaN(ts)) {
        reasons.push("timestamp is not parseable");
      }
    }

    if (reasons.length > 0) {
      rejected.push({ event, reasons });
    } else {
      accepted.push(event);
    }
  }

  ctx.accepted = accepted;
  ctx.rejected = rejected;

  return {
    accepted: accepted.length,
    rejected: rejected.length,
    rejection_reasons: rejected.map((r) => ({
      sensor_id: r.event.sensor_id ?? "unknown",
      reasons: r.reasons,
    })),
  };
}

/**
 * Step 3: Deduplicate — by sensor_id::timestamp key. First occurrence wins.
 */
function deduplicate(ctx) {
  const seen = new Set();
  const unique = [];
  let duplicateCount = 0;

  for (const event of ctx.accepted) {
    const key = `${event.sensor_id}::${event.timestamp}`;
    if (seen.has(key)) {
      duplicateCount++;
      continue;
    }
    seen.add(key);
    unique.push(event);
  }

  ctx.accepted = unique;

  return {
    before: unique.length + duplicateCount,
    after: unique.length,
    duplicates_removed: duplicateCount,
  };
}

/**
 * Step 4: Aggregate — hourly aggregates (sensor_id + hour bucket)
 * and per-location summary.
 */
function aggregate(ctx) {
  // Hourly aggregates: group by sensor_id + hour bucket
  /** @type {Map<string, {sensor_id: string, hour: string, type: string, unit: string, location: string, values: number[]}>} */
  const hourlyMap = new Map();

  for (const event of ctx.accepted) {
    const dt = new Date(event.timestamp);
    const hour = `${dt.getUTCFullYear()}-${String(dt.getUTCMonth() + 1).padStart(2, "0")}-${String(dt.getUTCDate()).padStart(2, "0")}T${String(dt.getUTCHours()).padStart(2, "0")}:00:00Z`;
    const key = `${event.sensor_id}::${hour}`;

    if (!hourlyMap.has(key)) {
      hourlyMap.set(key, {
        sensor_id: event.sensor_id,
        hour,
        type: event.type,
        unit: event.unit,
        location: event.location,
        values: [],
      });
    }
    hourlyMap.get(key).values.push(event.value);
  }

  const hourlyAggregates = [];
  for (const bucket of hourlyMap.values()) {
    const vals = bucket.values;
    const sum = vals.reduce((a, b) => a + b, 0);
    hourlyAggregates.push({
      sensor_id: bucket.sensor_id,
      hour: bucket.hour,
      type: bucket.type,
      unit: bucket.unit,
      location: bucket.location,
      count: vals.length,
      min: Math.min(...vals),
      max: Math.max(...vals),
      avg: Math.round((sum / vals.length) * 100) / 100,
      sum: Math.round(sum * 100) / 100,
    });
  }

  // Per-location summary
  /** @type {Map<string, {location: string, event_count: number, sensors: Set<string>, types: Set<string>}>} */
  const locationMap = new Map();

  for (const event of ctx.accepted) {
    if (!locationMap.has(event.location)) {
      locationMap.set(event.location, {
        location: event.location,
        event_count: 0,
        sensors: new Set(),
        types: new Set(),
      });
    }
    const loc = locationMap.get(event.location);
    loc.event_count++;
    loc.sensors.add(event.sensor_id);
    loc.types.add(event.type);
  }

  const locationSummary = [];
  for (const loc of locationMap.values()) {
    locationSummary.push({
      location: loc.location,
      event_count: loc.event_count,
      sensor_count: loc.sensors.size,
      sensors: [...loc.sensors].sort().join(";"),
      type_count: loc.types.size,
      types: [...loc.types].sort().join(";"),
    });
  }

  ctx.hourly_aggregates = hourlyAggregates;
  ctx.location_summary = locationSummary;

  return {
    hourly_buckets: hourlyAggregates.length,
    locations: locationSummary.length,
    sensors: new Set(ctx.accepted.map((e) => e.sensor_id)).size,
  };
}

/**
 * Step 5: Output — write accepted.jsonl, dead_letter.jsonl,
 * hourly_aggregates.csv, location_summary.csv, summary.json.
 */
async function output(ctx) {
  const outDir = ctx.outputDir;
  await mkdir(outDir, { recursive: true });

  const files = [];

  // accepted.jsonl
  const acceptedPath = join(outDir, "accepted.jsonl");
  const acceptedLines = ctx.accepted.map((e) => JSON.stringify(e));
  await writeFile(acceptedPath, acceptedLines.join("\n") + "\n", "utf-8");
  files.push(acceptedPath);

  // dead_letter.jsonl
  const deadLetterPath = join(outDir, "dead_letter.jsonl");
  const deadLetterLines = ctx.rejected.map((r) =>
    JSON.stringify({ event: r.event, reasons: r.reasons }),
  );
  const deadLetterContent =
    deadLetterLines.length > 0 ? deadLetterLines.join("\n") + "\n" : "";
  await writeFile(deadLetterPath, deadLetterContent, "utf-8");
  files.push(deadLetterPath);

  // hourly_aggregates.csv
  const aggPath = join(outDir, "hourly_aggregates.csv");
  const aggHeaders = [
    "sensor_id",
    "hour",
    "type",
    "unit",
    "location",
    "count",
    "min",
    "max",
    "avg",
    "sum",
  ];
  const aggRows = ctx.hourly_aggregates.map((row) =>
    aggHeaders.map((h) => row[h]).join(","),
  );
  await writeFile(
    aggPath,
    aggHeaders.join(",") + "\n" + aggRows.join("\n") + "\n",
    "utf-8",
  );
  files.push(aggPath);

  // location_summary.csv
  const locPath = join(outDir, "location_summary.csv");
  const locHeaders = [
    "location",
    "event_count",
    "sensor_count",
    "sensors",
    "type_count",
    "types",
  ];
  const locRows = ctx.location_summary.map((row) =>
    locHeaders.map((h) => row[h]).join(","),
  );
  await writeFile(
    locPath,
    locHeaders.join(",") + "\n" + locRows.join("\n") + "\n",
    "utf-8",
  );
  files.push(locPath);

  // summary.json
  const summaryPath = join(outDir, "summary.json");
  const summary = {
    total_events_read: ctx.raw_events.length,
    accepted: ctx.accepted.length,
    rejected: ctx.rejected.length,
    duplicates_removed: ctx.step_results.deduplicate.duplicates_removed,
    hourly_buckets: ctx.hourly_aggregates.length,
    locations: ctx.location_summary.length,
    sensors: [...new Set(ctx.accepted.map((e) => e.sensor_id))].sort(),
  };
  await writeFile(summaryPath, JSON.stringify(summary, null, 2) + "\n", "utf-8");
  files.push(summaryPath);

  ctx.output_files = files;

  return {
    files_written: files.length,
    output_dir: outDir,
  };
}

// ---------------------------------------------------------------------------
// Main pipeline
// ---------------------------------------------------------------------------

/**
 * Run the sensor demo pipeline end-to-end.
 *
 * @param {object} [options]
 * @param {string} [options.dataDir="data/sample"]
 * @param {string} [options.outputDir="data/silver/sensor_demo"]
 * @param {string} [options.manifestDir="data/manifests"]
 * @returns {Promise<{ pipeline_result: object, output_dir: string, manifest_path: string }>}
 */
export async function runSensorDemo({
  dataDir = "data/sample",
  outputDir = "data/silver/sensor_demo",
  manifestDir = "data/manifests",
} = {}) {
  const pipeline = new Pipeline("sensor_demo");

  pipeline
    .addStep("ingest", ingest)
    .addStep("validate", validate)
    .addStep("deduplicate", deduplicate)
    .addStep("aggregate", aggregate)
    .addStep("output", output);

  const context = { dataDir, outputDir };
  const result = await pipeline.run(context);

  // Write manifest after pipeline completes
  const runId = generateRunId();
  const manifestPath = writeManifest({
    pipeline_name: "sensor_demo",
    run_id: runId,
    source: join(dataDir, "sensor_events.json"),
    output: context.output_files ?? [],
    row_count: context.accepted ? context.accepted.length : 0,
    status: result.status,
    warnings: result.steps
      .filter((s) => s.status !== "success")
      .map((s) => `${s.name}: ${s.error}`),
    extras: {
      total_events_read: context.raw_events ? context.raw_events.length : 0,
      rejected_count: context.rejected ? context.rejected.length : 0,
      duplicates_removed: result.steps.find((s) => s.name === "deduplicate")
        ?.result?.duplicates_removed ?? 0,
    },
    manifestDir,
  });

  return {
    pipeline_result: result,
    output_dir: outputDir,
    manifest_path: manifestPath,
  };
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

async function main() {
  const { values } = parseArgs({
    options: {
      "data-dir": { type: "string", default: "../data/sample" },
      "output-dir": { type: "string", default: "../data/silver/sensor_demo" },
      "manifest-dir": { type: "string", default: "../data/manifests" },
      help: { type: "boolean", short: "h", default: false },
    },
    strict: true,
  });

  if (values.help) {
    console.log(
      `Usage: node src/sensor-demo.js [options]

Options:
  --data-dir <path>      Directory with sensor_events.json (default: ../data/sample)
  --output-dir <path>    Directory for pipeline output (default: ../data/silver/sensor_demo)
  --manifest-dir <path>  Directory for run manifest (default: ../data/manifests)
  -h, --help             Show this help message`,
    );
    process.exit(0);
  }

  const result = await runSensorDemo({
    dataDir: values["data-dir"],
    outputDir: values["output-dir"],
    manifestDir: values["manifest-dir"],
  });

  console.log();
  console.log(formatResult(result.pipeline_result));
  console.log();
  console.log(`Output directory: ${result.output_dir}`);
  console.log(`Manifest: ${result.manifest_path}`);
}

// Only run CLI when executed directly (not when imported)
const isMain = process.argv[1] && new URL(process.argv[1], "file://").pathname
  === new URL(import.meta.url).pathname;

if (isMain) {
  main().catch((err) => {
    console.error(err);
    process.exitCode = 1;
  });
}
